src/Stateless/TransitionConfiguration.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+
+namespace Stateless
+{
+    partial class StateMachine<TState, TTrigger>
+    {
+        /// <summary>
+        /// This class containd the required information for a transition.
+        /// </summary>
+        public class TransitionConfiguration
+        {
+            internal StateConfiguration StateConfiguration { get; private set; }
+
+            private readonly StateRepresentation _representation;
+            private readonly TTrigger _trigger;
+
+            /// <summary>
+            /// The TransitionConfiguration contains the information required to create a new transition.
+            /// </summary>
+            /// <param name="stateConfiguration"> Current state being configured</param>
+            /// <param name="representation">A reporesentation of the state</param>
+            /// <param name="trigger">Trigger for this transition</param>
+            internal TransitionConfiguration(StateConfiguration stateConfiguration, StateRepresentation representation, TTrigger trigger)
+            {
+                StateConfiguration = stateConfiguration;
+                _representation = representation;
+                _trigger = trigger;
+            }
+
+            /// <summary>
+            /// Adds a new transition to the destination state
+            /// </summary>
+            /// <param name="destination">Destination state</param>
+            public DestinationConfiguration To(TState destination)
+            {
+                TriggerBehaviour triggerBehaviour = new TransitioningTriggerBehaviour(_trigger, destination, null);
+                _representation.AddTriggerBehaviour(triggerBehaviour);
+                return new DestinationConfiguration(this, triggerBehaviour, _representation);
+            }
+
+            /// <summary>
+            /// Creates a new re-entrant transition. This transition, when triggered, will execute the state's
+            /// Exit and Entry actions, if any has been added.
+            /// </summary>
+            public DestinationConfiguration Self()
+            {
+                var destinationState = StateConfiguration.State;
+                var ttb = new TransitioningTriggerBehaviour(_trigger, destinationState, null);
+                _representation.AddTriggerBehaviour(ttb);
+                return new DestinationConfiguration(this, ttb, _representation);
+            }
+
+            /// <summary>
+            /// Creates a new internal transition. This transition, when triggered, not cause any state change,
+            /// nor will the Exit or Entry actions of the state be executed. Only the action configured will be
+            /// executed.
+            /// </summary>
+            /// <returns></returns>
+            public DestinationConfiguration Internal()
```

Comment:
There are times this would have been really useful on a previous project! However, is there some way that internal transitions could be represented differently in the dot graph to show that the entry & exit functions won't be executed?

Reply:
I guess so, someone would have to add it... In QM it looks like this:
![image](https://user-images.githubusercontent.com/13196280/79222559-0a1b3b00-7e58-11ea-80b1-89501f62dd99.png)

Reply:
This is QM, BTW:
https://www.state-machine.com/qm/

I've used QP for microcontroller state machines, and designed the state machines in QM.